### ISSUE TITLE
[ITensors] [BUG] op overloads blocked by other op method for "Qudit" sites

### ISSUE THREAD

**Issue by emstoudenmire** [ITensor/ITensors.jl#957]
**Description of bug**

Trying to overload `op` definitions for the "Qudit" SiteType using the style of `op` overload that just returns a matrix isn't working. Presumably this is because the generic version `op(on::OpName, st::SiteType"Qudit", s::Index...)` defined in qudit.jl is being called first, then trying to call functions named `_op` but not finding one with the custom operator name. (It is based on a similar bug report in the forum here: https://itensor.discourse.group/t/extending-op-function-definitions-for-the-qudit-sitetype/289)

**Minimal code demonstrating the bug or unexpected behavior**

Here is a sample code to reproduce:

<details><summary>Minimal runnable code</summary><p>

```julia
using ITensors
ITensors.op(::OpName"R",::SiteType"Qudit") =
 [1 0 0
  0 1 0
  0 0 1]
s = siteind("Qudit"; dim=3)
R = op("R",s)
```
</p></details>

**Expected output or behavior**

Expected the code to run without error and return an ITensor.

**Actual output or behavior**

<details><summary>Output of minimal runnable code</summary><p>

```julia
ERROR: MethodError: no method matching _op(::OpName{:R}, ::SiteType{Qudit}; dim=(3,))
Closest candidates are:
  _op(::OpName{:Id}, ::SiteType{Qudit}; dim) at ~/.julia/dev/ITensors/src/physics/site_types/qudit.jl:37
  _op(::OpName{:I}, ::SiteType{Qudit}; kwargs...) at ~/.julia/dev/ITensors/src/physics/site_types/qudit.jl:46
  _op(::OpName{:Adag}, ::SiteType{Qudit}; dim) at ~/.julia/dev/ITensors/src/physics/site_types/qudit.jl:50
  ...
Stacktrace:
 [1] op(on::OpName{:R}, st::SiteType{Qudit}, s::Index{Int64})
   @ ITensors ~/.julia/dev/ITensors/src/physics/site_types/qudit.jl:102
 [2] op(name::String, s::Index{Int64}; adjoint::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ITensors ~/.julia/dev/ITensors/src/physics/sitetype.jl:289
 [3] op(name::String, s::Index{Int64})
   @ ITensors ~/.julia/dev/ITensors/src/physics/sitetype.jl:226
 [4] top-level scope
   @ REPL[4]:1
```
</p></details>

**Version information**

 - Output from `versioninfo()`:
```julia
julia> versioninfo()
Julia Version 1.7.2
Commit bf53498635 (2022-02-06 15:21 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.5.0)
  CPU: Apple M1 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, westmere)
Environment:
  JULIA_DIR = /Applications/Julia-1.7.app/Contents/Resources/julia
```
 - Output from `using Pkg; Pkg.status("ITensors")`:
```julia
julia> using Pkg; Pkg.status("ITensors")
      Status `~/.julia/environments/v1.7/Project.toml`
  [9136182c] ITensors v0.3.18 `~/.julia/dev/ITensors`
```
(Actually my version is dev'd to the latest commit as of July 27.)

**Comment by mtfishman** [comment]
Currently the way to do this would be to overload `ITensors._op`, which is just defined for the `"Qudit"` type to forward the dimensions of the site indices. The reason the design is different from other site types is because the `"Qudit"`/`"Boson"` type has to handle variable dimensions.

So for example this should work:
```julia
using ITensors
ITensors._op(::OpName"R",::SiteType"Qudit") =
 [1 0 0
  0 1 0
  0 0 1]
s = siteind("Qudit"; dim=3)
R = op("R",s)
```
However this isn't a great external interface so we should think about a more robust design.

I would propose an interface where users have to overload the following function:
```julia
ITensors.op(::OpName"R", ::SiteType"Qudit", dims::Tuple) =
 [1 0 0
  0 1 0
  0 0 1]
```
where `dims` is a tuple of integers storing the dimensions of the sites that the operator acts on (i.e. a two-site operator acting on site indices of dimension `4` and `5` would be passed `(4, 5)`). The reason for having the extra `Tuple` input is to help dispatch distinguish from the version of `ITensors.op` that accepts a set of `Index`.

The example you show above is a bit pathological since in general you would expect an operator for `"Qudit"`/`"Boson"` to be defined for general on-site dimensions, but if it really is only defined for some fixed dimension I think it should be left to the user to check that, for example:
```julia
function ITensors.op(::OpName"R",::SiteType"Qudit", dims::Tuple{Int})
 @assert dims[1] == 3
 return [1 0 0
  0 1 0
  0 0 1]
end
```
Of course it would also get checked when the ITensor gets created, since it will check if the data being passed to the ITensor is consistent with the dimensions of the indices. Specifying the type as `Tuple{Int}` helps to add an extra check that the operator is only defined on a single site. You could imagine an operator like identity could be defined for any number of sites, so you could leave the final type as `dims::Tuple`.

**Comment by mtfishman** [comment]
This will be fixed by #963. A discussed offline with @emstoudenmire the new interface will be:
```julia
function ITensors.op(::OpName"my_qudit_op", ::SiteType"Qudit", d::Int)
  return [(i == j + 1) ? i * j : 0 for i in 1:d, j in 1:d]
end
d = 4
s = siteinds("Qudit", 2; dim=d)
o = op("my_qudit_op", s, 1)
@show o
```
outputs:
```julia
o = ITensor ord=2
Dim 1: (dim=4|id=951|"Qudit,Site,n=1")'
Dim 2: (dim=4|id=951|"Qudit,Site,n=1")
NDTensors.Dense{Float64, Vector{Float64}}
 4×4
 0.0  0.0   0.0  0.0
 2.0  0.0   0.0  0.0
 0.0  6.0   0.0  0.0
 0.0  0.0  12.0  0.0
```

**Comment by emstoudenmire** [comment]
Glad you got the int vararg version to work!
